File: compilers/ruby/test/test_n094_sighash_ir_wire_format.rb
# frozen_string_literal: true

# N-094 — the `--emit-ir` ANF wire format must carry `@sighash` losslessly.
#
# `--emit-ir` is a supported CLI mode and ANF is the declared conformance
# boundary, so ANF emitted by this tier is fed back into `--ir` (here, and by
# any other tier's loader) to produce a deployable locking script. A method
# declaring `/** @sighash SINGLE|FORKID */` compiles the BIP-143 flag byte
# 0x43 into the OP_PUSH_TX binding; when the flag is dropped from (or
# misspelled in) the emitted ANF the loader falls back to the default
# ALL|FORKID and the binding silently becomes 0x41. Same script length, one
# byte different — a covenant bound to the wrong sighash mode.
#
# Every assertion here is on the ROUND-TRIPPED SCRIPT HEX, not on the presence
# of a JSON key: the key is the mechanism, the byte is the harm.

require 'json'
require 'open3'
require 'set'
require 'tmpdir'
require_relative 'test_helper'
require 'runar_compiler/cli'

class TestN094SighashIRWireFormat < Minitest::Test
  RUBY_COMPILER_DIR = File.expand_path('..', __dir__)
  RUBY_CLI = File.join(RUBY_COMPILER_DIR, 'bin/runar-compiler-ruby')
  REPO_ROOT = File.expand_path('../..', RUBY_COMPILER_DIR)
  ANF_SCHEMA = File.join(REPO_ROOT, 'packages/runar-ir-schema/src/schemas/anf-ir.schema.json')
  GO_CLI = File.join(REPO_ROOT, 'compilers/go/runar-go')

  # SIGHASH_SINGLE | SIGHASH_FORKID.
  SIGHASH_SINGLE_FORKID = 0x43

  # Stateless contract with a manual checkPreimage under a non-default mode.
  SIGHASH_SRC = <<~TS
    import { SmartContract, SigHashPreimage, assert, checkPreimage } from "runar-lang";

    class C extends SmartContract {
      readonly s: bigint;

      constructor(s: bigint) { super(s); this.s = s; }

      /** @sighash SINGLE|FORKID */
      public m(flag: bigint, pre: SigHashPreimage): void {
        if (flag > 0n) {
          assert(checkPreimage(pre));
        } else {
          assert(flag <= 0n);
        }
        assert(this.s >= 0n);
      }
    }
  TS

  # Control: byte-for-byte the same contract with NO @sighash directive, so the
  # covenant runs under the default ALL|FORKID and the ANF carries no flag.
  DEFAULT_SRC = SIGHASH_SRC.sub("  /** @sighash SINGLE|FORKID */\n", '')

  # Control: no checkPreimage at all — nothing for the flag to ride on.
  NO_PREIMAGE_SRC = <<~TS
    import { SmartContract, assert } from "runar-lang";

    class C extends SmartContract {
      readonly s: bigint;

      constructor(s: bigint) { super(s); this.s = s; }

      public m(flag: bigint): void {
        assert(flag > 0n);
        assert(this.s >= 0n);
      }
    }
  TS

  def run_cli(*args)
    Open3.capture3('ruby', '-I', File.join(RUBY_COMPILER_DIR, 'lib'), RUBY_CLI, *args)
  end

  def with_source(src)
    Dir.mktmpdir('n094') do |dir|
      path = File.join(dir, 'C.runar.ts')
      File.write(path, src)
      yield dir, path
    end
  end

  # --source -> hex, --emit-ir -> --ir -> hex. Returns [direct_hex, round_hex, ir_json].
  def round_trip(src)
    with_source(src) do |dir, path|
      direct, err, st = run_cli('--source', path, '--hex')
      assert st.success?, "--source --hex failed: #{err}"

      ir, err, st = run_cli('--source', path, '--emit-ir')
      assert st.success?, "--emit-ir failed: #{err}"

      ir_path = File.join(dir, 'ir.json')
      File.write(ir_path, ir)
      round, err, st = run_cli('--ir', ir_path, '--hex')
      assert st.success?, "--ir --hex failed: #{err}"

      return [direct.strip, round.strip, ir]
    end
  end

  # Every `check_preimage` node anywhere in the program.
  def check_preimage_nodes(ir)
    found = []
    walk = lambda do |node|
      case node
      when Hash
        found << node if node['kind'] == 'check_preimage'
        node.each_value { |v| walk.call(v) }
      when Array
        node.each { |v| walk.call(v) }
      end
    end
    walk.call(JSON.parse(ir))
    found
  end

  # -------------------------------------------------------------------------
  # The harm: the round-tripped script byte
  # -------------------------------------------------------------------------

  def test_sighash_survives_the_emit_ir_round_trip
    direct, round, = round_trip(SIGHASH_SRC)

    refute_empty direct
    assert_equal direct, round,
                 "--emit-ir -> --ir changed the compiled script.\n" \
                 "The @sighash flag byte did not survive the ANF wire format:\n" \
                 "  direct: #{direct}\n  round : #{round}"
  end

  def test_round_tripped_script_still_commits_to_single_forkid
    direct, round, = round_trip(SIGHASH_SRC)
    default, = round_trip(DEFAULT_SRC)

    # Sanity: the two modes differ, and only in the flag byte, so an equality
    # assertion above is actually load-bearing (same length, one byte apart).
    assert_equal default.length, direct.length,
                 'expected SINGLE|FORKID and the default mode to be the same script length'
    refute_equal default, direct,
                 'expected @sighash SINGLE|FORKID to change the compiled script'
    refute_equal default, round,
                 'the round-tripped script silently downgraded to the DEFAULT sighash mode'
  end

  # -------------------------------------------------------------------------
  # Controls
  # -------------------------------------------------------------------------

  def test_control_default_mode_round_trips_byte_identically
    direct, round, ir = round_trip(DEFAULT_SRC)
    assert_equal direct, round, 'default-sighash contract no longer round-trips'
    check_preimage_nodes(ir).each do |n|
      refute n.key?('sighashFlag'),
             'default mode must omit sighashFlag so pre-existing goldens stay byte-identical'
    end
  end

  def test_control_contract_without_check_preimage_round_trips
    direct, round, ir = round_trip(NO_PREIMAGE_SRC)
    assert_equal direct, round, 'contract without checkPreimage no longer round-trips'
    assert_empty check_preimage_nodes(ir)
  end

  # -------------------------------------------------------------------------
  # Wire-format shape (the mechanism behind the byte)
  # -------------------------------------------------------------------------

  def test_check_preimage_carries_camelcase_sighash_flag
    _, _, ir = round_trip(SIGHASH_SRC)
    nodes = check_preimage_nodes(ir)
    assert_equal 1, nodes.size, 'expected exactly one check_preimage node'
    node = nodes.first

    refute node.key?('sighash_flag'),
           'snake_case sighash_flag is not the wire spelling; Go/Zig/Rust/Java read sighashFlag'
    assert_equal SIGHASH_SINGLE_FORKID, node['sighashFlag'],
                 "check_preimage lost the declared @sighash mode: #{node.inspect}"
  end

  def test_no_method_carries_a_stray_sighash_type
    _, _, ir = round_trip(SIGHASH_SRC)
    JSON.parse(ir)['methods'].each do |m|
      refute m.key?('sighash_type'),
             "method #{m['name']} leaks the in-memory sighash_type carrier into the ANF wire format"
      refute m.key?('sighashType'),
             "method #{m['name']} leaks the in-memory sighash_type carrier into the ANF wire format"
    end
  end

  # -------------------------------------------------------------------------
  # Cross-tier: the reported reproduction. Go loads our ANF.
  # -------------------------------------------------------------------------

  def test_go_compiles_our_anf_to_the_same_script
    skip "go compiler binary not built at #{GO_CLI}" unless File.executable?(GO_CLI)

    with_source(SIGHASH_SRC) do |dir, path|
      go_direct, err, st = Open3.capture3(GO_CLI, '--source', path, '--hex')
      assert st.success?, "go --source --hex failed: #{err}"

      ir, err, st = run_cli('--source', path, '--emit-ir')
      assert st.success?, "--emit-ir failed: #{err}"
      ir_path = File.join(dir, 'ir.json')
      File.write(ir_path, ir)

      go_round, err, st = Open3.capture3(GO_CLI, '--ir', ir_path, '--hex')
      assert st.success?, "go --ir --hex failed: #{err}"

      assert_equal go_direct.strip, go_round.strip,
                   'Go compiled a DIFFERENT script from our ANF than from the same source'
    end
  end

  # -------------------------------------------------------------------------
  # Lockstep guard — the class of defect, not this instance.
  #
  # The wire names the ANF JSON may use are fixed by the cross-tier JSON
  # Schema in packages/runar-ir-schema. Every field an ANF node type declares
  # must therefore either (a) serialize to a key that schema knows, and
  # actually appear in the serializer's output, or (b) be listed in the
  # serializer's documented exclusion set. A field that is neither is a field
  # the serializer silently drops or misspells — exactly what happened to
  # sighash_flag. The field list is read off the TYPE, so this fires the
  # moment the type gains a field the serializer does not handle.
  # -------------------------------------------------------------------------

  # Fields this tier emits under a name the ANF JSON Schema does not list.
  #
  # EMPTY, and it must stay that way: an entry here is a field whose bytes no
  # other tier can read. N-095 emptied it by settling `synthetic_array_chain`
  # on the Go spelling (`syntheticArrayChain`) and declaring it in
  # `$defs.ANFProperty`, so this guard -- not a comment -- is now what stops
  # the spelling from splitting three ways again.
  SCHEMA_DIVERGENCES = [].freeze

  # Wire keys the ANF JSON Schema accepts anywhere in the document.
  def schema_wire_keys
    schema = JSON.parse(File.read(ANF_SCHEMA))
    keys = Set.new(schema.fetch('properties').keys)
    schema.fetch('$defs').each_value do |d|
      keys.merge(d['properties'].keys) if d.is_a?(Hash) && d['properties'].is_a?(Hash)
    end
    keys
  end

  # A sentinel for each field, chosen so the serializer's non-nil test passes
  # and nested walks do not explode.
  def populate(obj, fields, kind)
    fields.each do |f|
      val = case f.to_s
            when 'kind' then kind
            when 'then', 'else_', 'body', 'args', 'state_values', 'elements',
                 'params', 'properties', 'methods', 'synthetic_array_chain' then []
            # Non-empty: the artifact ANF serializer skips an empty +results+.
            when 'results' then ['x']
            when 'count', 'start', 'step', 'in_arity', 'out_arity', 'sighash_flag',
                 'sighash_type', 'line', 'column' then 1
            when 'readonly', 'is_public', 'is_auto_injected_state_check', 'preserve' then true
            when 'value' then RunarCompiler::IR::ANFValue.new(kind: 'get_state_script')
            when 'source_loc' then RunarCompiler::IR::SourceLocation.new
            when 'raw_value' then '"x"'
            else 'x'
            end
      obj.respond_to?(:[]=) && obj.is_a?(Struct) ? obj[f] = val : obj.send("#{f}=", val)
    end
    obj
  end

  def assert_fields_are_all_accounted_for(node, fields, label, wire_keys)
    # A few fields are emitted only for one +kind+ (isAutoInjectedStateCheck
    # rides on +assert+, sighashFlag on +check_preimage+), so serialize the
    # fully-populated node once per discriminator and union the wire keys.
    emitted = %w[assert check_preimage].flat_map do |kind|
      RunarCompiler::CLI._anf_to_camel_dict(populate(node, fields, kind)).keys
    end.uniq
    excluded = RunarCompiler::CLI::IR_EXCLUDED_FIELDS

    fields.map(&:to_s).each do |f|
      next if excluded.include?(f)

      key = RunarCompiler::CLI._snake_key(f)
      unless SCHEMA_DIVERGENCES.include?(f)
        assert_includes wire_keys, key,
                        "#{label}.#{f} serializes to \"#{key}\", which the cross-tier ANF JSON " \
                        "Schema does not accept. Fix the wire name, or exclude the field."
      end
      assert_includes emitted, key,
                      "#{label}.#{f} never reaches the emitted ANF (expected key \"#{key}\"). " \
                      'Fields dropped here vanish silently: the Go loader ignores unknown keys.'
    end
  end

  def test_every_anf_field_is_either_emitted_or_explicitly_excluded
    wire_keys = schema_wire_keys
    ir = RunarCompiler::IR

    value = ir::ANFValue.new(kind: 'check_preimage')
    value_fields = value.instance_variables.map { |s| s.to_s.delete_prefix('@').to_sym }
    assert_includes value_fields, :sighash_flag, 'ANFValue no longer declares sighash_flag'
    assert_fields_are_all_accounted_for(value, value_fields, 'ANFValue', wire_keys)

    {
      'ANFProgram' => ir::ANFProgram.new,
      'ANFProperty' => ir::ANFProperty.new,
      'ANFMethod' => ir::ANFMethod.new,
      'ANFParam' => ir::ANFParam.new,
      'ANFBinding' => ir::ANFBinding.new,
    }.each do |label, node|
      assert_fields_are_all_accounted_for(node, node.members, label, wire_keys)
    end
  end

  # -------------------------------------------------------------------------
  # The SECOND ANF serializer in this tier.
  #
  # +Compiler::_serialize_anf_program+ fills the +anf+ field the artifact
  # carries for every STATEFUL contract -- the copy SDK ANF interpreters read.
  # It is a third hand-maintained enumeration of the same field list (the
  # emit allowlist and the rename table were the other two), so it gets the
  # same field-coverage guard.
  # -------------------------------------------------------------------------

  # Fully-populated ANFValue serialized through the artifact ANF serializer,
  # unioned over the discriminators that gate a field (+assert+ gates
  # isAutoInjectedStateCheck, +raw_script+ the byte span, and so on).
  def artifact_anf_value_keys(fields)
    ir = RunarCompiler::IR
    %w[assert check_preimage raw_script array_literal].flat_map do |kind|
      program = ir::ANFProgram.new(
        contract_name: 'C',
        properties: [],
        methods: [ir::ANFMethod.new(
          name: 'm',
          params: [],
          body: [ir::ANFBinding.new(
            name: 't0',
            value: populate(ir::ANFValue.new(kind: kind), fields, kind),
          )],
          is_public: true,
        )],
      )
      RunarCompiler.send(:_serialize_anf_program, program)['methods'][0]['body'][0]['value'].keys
    end.uniq
  end

  def test_artifact_anf_serializer_emits_every_declared_field
    wire_keys = schema_wire_keys
    fields = RunarCompiler::IR::ANFValue::FIELDS
    emitted = artifact_anf_value_keys(fields)
    excluded = RunarCompiler::CLI::IR_EXCLUDED_FIELDS

    fields.map(&:to_s).each do |f|
      next if excluded.include?(f)

      key = RunarCompiler::CLI._snake_key(f)
      assert_includes emitted, key,
                      "ANFValue.#{f} never reaches the artifact's embedded ANF " \
                      "(expected key \"#{key}\"). Stateful artifacts carry that ANF to the " \
                      'SDK interpreters, so a field dropped here is a field they cannot see.'
      assert_includes wire_keys, key, "unexpected wire key \"#{key}\" for ANFValue.#{f}"
    end
  end

  # The concrete harm behind the guard above: a stateful contract whose body
  # builds an array (checkMultiSig) must carry the array's element refs.
  MULTISIG_STATEFUL_SRC = <<~TS
    import { StatefulSmartContract, assert, checkMultiSig, PubKey, Sig } from "runar-lang";

    class Arr2 extends StatefulSmartContract {
      n: bigint;
      readonly a: PubKey;
      readonly b: PubKey;

      constructor(n: bigint, a: PubKey, b: PubKey) { super(n, a, b); this.n = n; this.a = a; this.b = b; }

      public bump(s1: Sig, s2: Sig): void {
        assert(checkMultiSig([s1, s2], [this.a, this.b]));
        this.n = this.n + 1n;
      }
    }
  TS

  def test_stateful_artifact_anf_keeps_array_literal_elements
    with_source(MULTISIG_STATEFUL_SRC) do |_dir, path|
      out, err, st = run_cli('--source', path)
      assert st.success?, "--source failed: #{err}"

      artifact = JSON.parse(out)
      anf = artifact['anf']
      refute_nil anf, 'a stateful contract must carry its ANF in the artifact'

      nodes = []
      walk = lambda do |n|
        case n
        when Hash
          nodes << n if n['kind'] == 'array_literal'
          n.each_value { |v| walk.call(v) }
        when Array then n.each { |v| walk.call(v) }
        end
      end
      walk.call(anf)

      refute_empty nodes, 'expected checkMultiSig to lower to array_literal nodes'
      nodes.each do |n|
        refute_nil n['elements'],
                   "array_literal in the artifact ANF has no elements: #{n.inspect}. " \
                   'The Go reference emits {"kind":"array_literal","elements":[...]}; without ' \
                   'them an SDK ANF interpreter cannot rebuild the array.'
        refute_empty n['elements']
      end
    end
  end
end

File: compilers/python/tests/test_n094_sighash_ir_wire_format.py
"""N-094 — the ``--emit-ir`` ANF wire format must carry ``@sighash`` losslessly.

``--emit-ir`` is a supported CLI mode and ANF is the declared conformance
boundary, so ANF emitted by this tier is fed back into ``--ir`` (here, and by
any other tier's loader) to produce a deployable locking script. A method
declaring ``/** @sighash SINGLE|FORKID */`` compiles the BIP-143 flag byte
0x43 into the OP_PUSH_TX binding; when the flag is dropped from (or misspelled
in) the emitted ANF the loader falls back to the default ALL|FORKID and the
binding silently becomes 0x41. Same script length, one byte different — a
covenant bound to the wrong sighash mode.

Every assertion here is on the ROUND-TRIPPED SCRIPT HEX, not on the presence of
a JSON key: the key is the mechanism, the byte is the harm.
"""

from __future__ import annotations

import dataclasses
import json
import os
import subprocess
import sys
from pathlib import Path

import pytest

from runar_compiler import __main__ as cli_main
from runar_compiler.ir import types as ir_types

PYTHON_COMPILER_DIR = Path(__file__).resolve().parent.parent
REPO_ROOT = PYTHON_COMPILER_DIR.parent.parent
ANF_SCHEMA = REPO_ROOT / "packages" / "runar-ir-schema" / "src" / "schemas" / "anf-ir.schema.json"
GO_CLI = REPO_ROOT / "compilers" / "go" / "runar-go"

#: SIGHASH_SINGLE | SIGHASH_FORKID.
SIGHASH_SINGLE_FORKID = 0x43

#: Stateless contract with a manual checkPreimage under a non-default mode.
SIGHASH_SRC = """\
import { SmartContract, SigHashPreimage, assert, checkPreimage } from "runar-lang";

class C extends SmartContract {
  readonly s: bigint;

  constructor(s: bigint) { super(s); this.s = s; }

  /** @sighash SINGLE|FORKID */
  public m(flag: bigint, pre: SigHashPreimage): void {
    if (flag > 0n) {
      assert(checkPreimage(pre));
    } else {
      assert(flag <= 0n);
    }
    assert(this.s >= 0n);
  }
}
"""

#: Control: the same contract with NO @sighash directive — default ALL|FORKID.
DEFAULT_SRC = SIGHASH_SRC.replace("  /** @sighash SINGLE|FORKID */\n", "")

#: Control: no checkPreimage at all — nothing for the flag to ride on.
NO_PREIMAGE_SRC = """\
import { SmartContract, assert } from "runar-lang";

class C extends SmartContract {
  readonly s: bigint;

  constructor(s: bigint) { super(s); this.s = s; }

  public m(flag: bigint): void {
    assert(flag > 0n);
    assert(this.s >= 0n);
  }
}
"""


def _run(args: list[str]) -> subprocess.CompletedProcess:
    return subprocess.run(
        [sys.executable, "-m", "runar_compiler", *args],
        cwd=str(PYTHON_COMPILER_DIR),
        capture_output=True,
        text=True,
        timeout=120,
        env=os.environ.copy(),
    )


def _round_trip(tmp_path: Path, src: str) -> tuple[str, str, dict]:
    """``--source --hex`` vs ``--emit-ir`` -> ``--ir --hex``.

    Returns ``(direct_hex, round_tripped_hex, emitted_anf)``.
    """
    source = tmp_path / "C.runar.ts"
    source.write_text(src)

    direct = _run(["--source", str(source), "--hex"])
    assert direct.returncode == 0, f"--source --hex failed: {direct.stderr}"

    emitted = _run(["--source", str(source), "--emit-ir"])
    assert emitted.returncode == 0, f"--emit-ir failed: {emitted.stderr}"

    ir_path = tmp_path / "ir.json"
    ir_path.write_text(emitted.stdout)

    rounded = _run(["--ir", str(ir_path), "--hex"])
    assert rounded.returncode == 0, f"--ir --hex failed: {rounded.stderr}"

    return direct.stdout.strip(), rounded.stdout.strip(), json.loads(emitted.stdout)


def _check_preimage_nodes(anf: object) -> list[dict]:
    found: list[dict] = []
    if isinstance(anf, dict):
        if anf.get("kind") == "check_preimage":
            found.append(anf)
        for v in anf.values():
            found.extend(_check_preimage_nodes(v))
    elif isinstance(anf, list):
        for v in anf:
            found.extend(_check_preimage_nodes(v))
    return found


# ---------------------------------------------------------------------------
# The harm: the round-tripped script byte
# ---------------------------------------------------------------------------


def test_sighash_survives_the_emit_ir_round_trip(tmp_path: Path) -> None:
    direct, rounded, _ = _round_trip(tmp_path, SIGHASH_SRC)
    assert direct
    assert direct == rounded, (
        "--emit-ir -> --ir changed the compiled script. The @sighash flag byte "
        f"did not survive the ANF wire format:\n  direct: {direct}\n  round : {rounded}"
    )


def test_round_tripped_script_still_commits_to_single_forkid(tmp_path: Path) -> None:
    single_dir = tmp_path / "single"
    single_dir.mkdir()
    default_dir = tmp_path / "default"
    default_dir.mkdir()
    direct, rounded, _ = _round_trip(single_dir, SIGHASH_SRC)
    default, _, _ = _round_trip(default_dir, DEFAULT_SRC)

    # Same length, one byte apart — which is exactly why the equality assertion
    # above is load-bearing and a length check would not be.
    assert len(default) == len(direct)
    assert default != direct, "expected @sighash SINGLE|FORKID to change the compiled script"
    assert default != rounded, (
        "the round-tripped script silently downgraded to the DEFAULT sighash mode"
    )


# ---------------------------------------------------------------------------
# Controls
# ---------------------------------------------------------------------------


def test_control_default_mode_round_trips_byte_identically(tmp_path: Path) -> None:
    direct, rounded, anf = _round_trip(tmp_path, DEFAULT_SRC)
    assert direct == rounded, "default-sighash contract no longer round-trips"
    for node in _check_preimage_nodes(anf):
        assert "sighashFlag" not in node, (
            "default mode must omit sighashFlag so pre-existing goldens stay byte-identical"
        )


def test_control_contract_without_check_preimage_round_trips(tmp_path: Path) -> None:
    direct, rounded, anf = _round_trip(tmp_path, NO_PREIMAGE_SRC)
    assert direct == rounded, "contract without checkPreimage no longer round-trips"
    assert _check_preimage_nodes(anf) == []


# ---------------------------------------------------------------------------
# Wire-format shape (the mechanism behind the byte)
# ---------------------------------------------------------------------------


def test_check_preimage_carries_camelcase_sighash_flag(tmp_path: Path) -> None:
    _, _, anf = _round_trip(tmp_path, SIGHASH_SRC)
    nodes = _check_preimage_nodes(anf)
    assert len(nodes) == 1, "expected exactly one check_preimage node"
    node = nodes[0]

    assert "sighash_flag" not in node, (
        "snake_case sighash_flag is not the wire spelling; Go/Zig/Rust/Java read sighashFlag"
    )
    assert node.get("sighashFlag") == SIGHASH_SINGLE_FORKID, (
        f"check_preimage lost the declared @sighash mode: {node!r}"
    )


def test_no_method_carries_a_stray_sighash_type(tmp_path: Path) -> None:
    _, _, anf = _round_trip(tmp_path, SIGHASH_SRC)
    for method in anf["methods"]:
        for key in ("sighash_type", "sighashType"):
            assert key not in method, (
                f"method {method['name']} leaks the in-memory sighash_type carrier "
                "into the ANF wire format"
            )


# ---------------------------------------------------------------------------
# Cross-tier: the reported reproduction. Go loads our ANF.
# ---------------------------------------------------------------------------


def test_go_compiles_our_anf_to_the_same_script(tmp_path: Path) -> None:
    if not (GO_CLI.exists() and os.access(GO_CLI, os.X_OK)):
        pytest.skip(f"go compiler binary not built at {GO_CLI}")

    source = tmp_path / "C.runar.ts"
    source.write_text(SIGHASH_SRC)

    go_direct = subprocess.run(
        [str(GO_CLI), "--source", str(source), "--hex"],
        capture_output=True, text=True, timeout=120,
    )
    assert go_direct.returncode == 0, f"go --source --hex failed: {go_direct.stderr}"

    emitted = _run(["--source", str(source), "--emit-ir"])
    assert emitted.returncode == 0, f"--emit-ir failed: {emitted.stderr}"
    ir_path = tmp_path / "ir.json"
    ir_path.write_text(emitted.stdout)

    go_round = subprocess.run(
        [str(GO_CLI), "--ir", str(ir_path), "--hex"],
        capture_output=True, text=True, timeout=120,
    )
    assert go_round.returncode == 0, f"go --ir --hex failed: {go_round.stderr}"

    assert go_direct.stdout.strip() == go_round.stdout.strip(), (
        "Go compiled a DIFFERENT script from our ANF than from the same source"
    )


# ---------------------------------------------------------------------------
# Lockstep guard — the class of defect, not this instance.
#
# The wire names the ANF JSON may use are fixed by the cross-tier JSON Schema in
# packages/runar-ir-schema. Every field an ANF node type declares must therefore
# either (a) serialize to a key that schema knows, and actually appear in the
# serializer's output, or (b) be listed in the serializer's documented exclusion
# set. A field that is neither is a field the serializer silently drops or
# misspells — exactly what happened to sighash_flag. The field list is read off
# the TYPE, so this fires the moment the type gains a field the serializer does
# not handle.
# ---------------------------------------------------------------------------


def _schema_wire_keys() -> set[str]:
    schema = json.loads(ANF_SCHEMA.read_text())
    keys = set(schema["properties"])
    for definition in schema["$defs"].values():
        if isinstance(definition, dict) and isinstance(definition.get("properties"), dict):
            keys |= set(definition["properties"])
    return keys


def _sentinel(name: str, kind: str) -> object:
    if name == "kind":
        return kind
    if name in {
        "then", "else_", "body", "args", "results", "state_values", "elements",
        "params", "properties", "methods",
    }:
        return []
    # N-095: this one must be NON-empty. The emitter skips an empty chain (as
    # Go's `omitempty` and Rust's `skip_serializing_if` do), so an `[]`
    # sentinel would make the "field reaches the wire" half of this guard
    # vacuous — exactly the blind spot that let three tiers spell the key three
    # different ways.
    if name == "synthetic_array_chain":
        return [{"base": "grid", "index": 0, "length": 2}]
    if name in {"count", "start", "step", "in_arity", "out_arity", "sighash_flag",
                "sighash_type", "line", "column"}:
        return 1
    if name in {"readonly", "is_public", "is_auto_injected_state_check", "preserve"}:
        return True
    if name == "value":
        return ir_types.ANFValue(kind="get_state_script")
    if name == "source_loc":
        return ir_types.SourceLocation()
    if name == "raw_value":
        return '"x"'
    return "x"


def _assert_fields_accounted_for(node_type: type, label: str, wire_keys: set[str]) -> None:
    fields = [f.name for f in dataclasses.fields(node_type)]

    # A few fields are emitted only for one `kind` (isAutoInjectedStateCheck
    # rides on `assert`, sighashFlag on `check_preimage`), so serialize the
    # fully-populated node once per discriminator and union the wire keys.
    emitted: set[str] = set()
    for kind in ("assert", "check_preimage"):
        node = node_type(**{name: _sentinel(name, kind) for name in fields})
        out = cli_main._anf_to_camel_dict(node)
        assert isinstance(out, dict)
        emitted |= set(out)

    for name in fields:
        if name in cli_main._IR_EXCLUDED_FIELDS:
            continue
        key = cli_main._snake_key(name)
        assert key in wire_keys, (
            f'{label}.{name} serializes to "{key}", which the cross-tier ANF JSON Schema '
            "does not accept. Fix the wire name, or exclude the field."
        )
        assert key in emitted, (
            f'{label}.{name} never reaches the emitted ANF (expected key "{key}"). '
            "Fields dropped here vanish silently: the Go loader ignores unknown keys."
        )


def test_every_anf_field_is_either_emitted_or_explicitly_excluded() -> None:
    wire_keys = _schema_wire_keys()

    declared = {f.name for f in dataclasses.fields(ir_types.ANFValue)}
    assert "sighash_flag" in declared, "ANFValue no longer declares sighash_flag"

    for node_type, label in (
        (ir_types.ANFValue, "ANFValue"),
        (ir_types.ANFProgram, "ANFProgram"),
        (ir_types.ANFProperty, "ANFProperty"),
        (ir_types.ANFMethod, "ANFMethod"),
        (ir_types.ANFParam, "ANFParam"),
        (ir_types.ANFBinding, "ANFBinding"),
    ):
        _assert_fields_accounted_for(node_type, label, wire_keys)
